lib/auth.js

```diff
@@ -0,0 +1,248 @@
+'use strict';
+
+const unicodeCategories = require('./utils/unicode-categories');
+
+const unicodeRangesIncludes = (ranges, codePoint) => {
+  let left = 0;
+  let right = ranges.length - 1;
+  while (left <= right) {
+    const mid = (left + right) >>> 1;
+    const value = ranges[mid];
+    if (typeof value === 'number') {
+      if (codePoint > value) left = mid + 1;
+      else if (codePoint < value) right = mid - 1;
+      else return true;
+    } else if (codePoint > value[1]) {
+      left = mid + 1;
+    } else if (codePoint < value[0]) {
+      right = mid - 1;
+    } else {
+      return true;
+    }
+  }
+  return false;
+};
+
+const stringIncludesChars = (str, ranges, charsNumber) => {
+  let number = 0;
+  const utf16singleUnit = 1 << 16;
+  for (let index = 0; index <= str.length - (charsNumber - number); index++) {
+    const codePoint = str.codePointAt(index);
+    if (codePoint > utf16singleUnit) index++;
+    if (unicodeRangesIncludes(ranges, codePoint)) {
+      if (++number === charsNumber) return true;
+    }
+  }
+  return false;
+};
+
+const passwordTests = {
+  MIN_LENGTH: {
+    test: (password, options) => password.length >= options.minLength,
+    hint: options => 'The password must be longer than or equal to ' +
```

Comment:
I think wording it like `'password must be at least N chars long'` will be better.
Also, AFAIU, these hints will also be shown to the end-user; therefore we have to somehow integrate i18n for the hints to allow them to be localized when needed.